crypto/sig_abstractions.go

```diff
@@ -0,0 +1,166 @@
+	AlgorithmType uint64
+)
+
+// all AlgorithmType enums
+const (
+	minAlgorithmType AlgorithmType = iota
+
+	DilithiumType
+	Ed25519Type
+
+	maxAlgorithmType
+)
+
+// Signer interface represents the possible operations that can be done with a signing key.
+type Signer interface {
+	Sign(message Hashable) ByteSignature
+	SignBytes(message []byte) ByteSignature
+	GetVerifyingKey() *VerifyingKey
+}
+
+// ErrBadSignature represents a bad signature
+var ErrBadSignature = fmt.Errorf("invalid signature")
+
+// Verifier interface represents any algorithm that can verify signatures for a specific signing scheme.
+type Verifier interface {
+	// Verify and VerifyBytes returns error on bad signature, and any other problem.
+	Verify(message Hashable, sig ByteSignature) error
+	VerifyBytes(message []byte, sig ByteSignature) error
+}
+
+// SignatureAlgorithm holds a Signer, and the type of algorithm the Signer conforms with.
+// to add a key - verify that PackedSignatureAlgorithm's function (getSigner) returns your key.
+type SignatureAlgorithm struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Type AlgorithmType            `codec:"sigType"`
+	Pack PackedSignatureAlgorithm `codec:"keys"`
+}
+
+// VerifyingKey is an abstraction of a key store of verifying keys.
+// it can return the correct key according to the underlying algorithm.
+// Implements Hashable too.
+//
+// NOTE: The VerifyingKey key might not be a valid key if a malicious client sent it over the network
+// make certain it is valid.
+type VerifyingKey struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Type AlgorithmType      `codec:"type"`
+	Pack PackedVerifyingKey `codec:"pks"`
+}
+
+// ToBeHashed makes it easier to hash the VeryfyingKey struct.
+func (z *VerifyingKey) ToBeHashed() (protocol.HashID, []byte) {
+	return protocol.VerifyingKey, protocol.Encode(z)
+}
+
+// GetSigner fetches the Signer type that is stored inside this SignatureAlgorithm.
+func (z *SignatureAlgorithm) GetSigner() (Signer, error) {
+	return z.Pack.getSigner(z.Type)
+}
+
+// GetVerifier fetches the Verifier type that is stored inside this VerifyingKey.
+func (z *VerifyingKey) GetVerifier() (Verifier, error) {
+	return z.Pack.getVerifier(z.Type)
+}
+
+// PackedVerifyingKey is a key store. Allows for easy marshal/unmarshal.
+type PackedVerifyingKey struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	DilithiumPublicKey DilithiumVerifier `codec:"dpk"`
+	Ed25519PublicKey   Ed25519PublicKey  `codec:"edpk"`
+}
+
+var errUnknownVerifier = errors.New("could not find stored Verifier")
+
+func (p *PackedVerifyingKey) getVerifier(t AlgorithmType) (Verifier, error) {
+	switch t {
+	case DilithiumType:
+		return &p.DilithiumPublicKey, nil
+	case Ed25519Type:
+		return &p.Ed25519PublicKey, nil
+	default:
+		return nil, errUnknownVerifier
+	}
+}
+
+// PackedSignatureAlgorithm helps  marshal SignatureAlgorithm
+type PackedSignatureAlgorithm struct {
```

Comment:
I don't see the benefit of the structs `PackedVerifyingKey` and `PackedSignatureAlgorithm`. I think it would be simpler if `VerifyingKey` and `SignatureAlgorithm` simply had the key fields directly in their structs, since it avoids an additional layer of indirection.